main/src/ca/uwaterloo/flix/language/phase/Verifier.scala

```diff
@@ -528,6 +565,78 @@ object Verifier {
+  }
+
+  /**
+    * Asserts that `tpe` is a subtype of the java class type `klazz`.
+    */
+  @tailrec
+  private def checkJavaSubtype(tpe: MonoType, klazz: Class[_], loc: SourceLocation): MonoType = {
+    val monoToPrimitiveName = Map[MonoType, String](
+      MonoType.Int8 -> "byte",
+      MonoType.Int16 -> "short",
+      MonoType.Int32 -> "int",
+      MonoType.Int64 -> "long",
+      MonoType.Float32 -> "float",
+      MonoType.Float64 -> "double",
+      MonoType.Bool -> "boolean",
+      MonoType.Char -> "char",
+      MonoType.Unit -> "void"
+    )
+
+    val monoToClassName = Map[MonoType, String](
+      MonoType.String -> "java.lang.String",
+      MonoType.BigInt -> "java.math.BigInteger",
+      MonoType.BigDecimal -> "java.math.BigDecimal",
+      MonoType.Regex -> "java.util.regex.Pattern",
+      MonoType.Arrow(List(MonoType.Object), MonoType.Unit) -> "java.util.function.Consumer",
+      MonoType.Arrow(List(MonoType.Object), MonoType.Bool) -> "java.util.function.Predicate",
+      MonoType.Arrow(List(MonoType.Int32), MonoType.Unit) -> "java.util.function.IntConsumer",
+      MonoType.Arrow(List(MonoType.Int32), MonoType.Object) -> "java.util.function.IntFunction",
+      MonoType.Arrow(List(MonoType.Int32), MonoType.Bool) -> "java.util.function.IntPredicate",
+      MonoType.Arrow(List(MonoType.Int32), MonoType.Int32) -> "java.util.function.IntUnaryOperator",
+      MonoType.Arrow(List(MonoType.Int32), MonoType.Unit) -> "java.util.function.IntConsumer",
+      MonoType.Arrow(List(MonoType.Int64), MonoType.Unit) -> "java.util.function.LongConsumer",
+      MonoType.Arrow(List(MonoType.Int64), MonoType.Object) -> "java.util.function.LongFunction",
+      MonoType.Arrow(List(MonoType.Int64), MonoType.Bool) -> "java.util.function.LongPredicate",
+      MonoType.Arrow(List(MonoType.Int64), MonoType.Int64) -> "java.util.function.LongUnaryOperator",
+      MonoType.Arrow(List(MonoType.Float64), MonoType.Unit) -> "java.util.function.DoubleConsumer",
+      MonoType.Arrow(List(MonoType.Float64), MonoType.Object) -> "java.util.function.DoubleFunction",
+      MonoType.Arrow(List(MonoType.Float64), MonoType.Bool) -> "java.util.function.DoublePredicate",
+      MonoType.Arrow(List(MonoType.Float64), MonoType.Float64) -> "java.util.function.DoubleUnaryOperator",
+    )
+
+    tpe match {
+      case MonoType.Array(elmt) if klazz.isArray =>
+        checkJavaSubtype(elmt, klazz.getComponentType, loc)
+
+      case MonoType.Native(k) if isSubclass(k, klazz) =>
+        tpe
+
+      case _ if monoToPrimitiveName.contains(tpe) && klazz.getName == monoToPrimitiveName(tpe) =>
+        tpe
+
+      case _ if monoToClassName.contains(tpe) && isSubclass(Class.forName(monoToClassName(tpe)), klazz) =>
+        tpe
+
+      case _ => failMismatchedTypes(tpe, klazz, loc)
+    }
+  }
+
+  /**
+    * Returns `true` if and only if `c1` is a subclass of `c2`
+    */
+  @tailrec
+  private def isSubclass(c1: Class[_], c2: Class[_]): Boolean =
```

Comment:
I think Class supports this inherently with `isAssignableFrom` or something like that, that also allows `int[] <: Object` and other weird cases